### ISSUE TITLE
Is this called memory leak ?

### ISSUE THREAD

**Issue by BabyfaceDeveloper** [dotnet/efcore#27876]
I've created a small asp.net core app to learn insert in efcore but I notice the issue about memory leak when I insert 1 million records, memory does not decrease after done inserting
Here the memory of my app before running
![image](https://user-images.githubusercontent.com/58478947/164983114-69d27b88-71e2-40cc-a215-79dc139984d1.png)

### Configuration

* Platform: .Net 6
* OS: Windows 10
* CPU architecture: x64
* Database: Postgres
* App configuration, see below:

1.  HomeController.cs
![image](https://user-images.githubusercontent.com/58478947/164982780-f61e2f6b-0a90-4904-b06e-9b225a52a429.png)
2.  Program.cs
![image](https://user-images.githubusercontent.com/58478947/164982820-23d7d18d-8ff8-4968-821a-0f8fe541f34f.png)
3. TestService.cs
![image](https://user-images.githubusercontent.com/58478947/164982852-61fdadc8-0765-4735-855c-a1b8b989d3c6.png)
4. AppDbContext.cs
![image](https://user-images.githubusercontent.com/58478947/164982881-d9dcdc20-9fc3-488a-b1a3-d63f7dde35ae.png)
5. Test.cs
![image](https://user-images.githubusercontent.com/58478947/164982937-b093b93f-86de-442b-995a-177392996f66.png)
6. demoDotnet.csproj
![image](https://user-images.githubusercontent.com/58478947/164982969-a8cf6088-5a03-4e74-83ae-f496a3900100.png) 
7. Test data
![image](https://user-images.githubusercontent.com/58478947/164983017-b09f62d6-c7c3-4eea-9950-26abae39861b.png)
8.  launchSetting.json
![image](https://user-images.githubusercontent.com/58478947/164983034-729ba0f4-63f3-4d30-9a28-fda7cae5e153.png)

### Results

* Tool use for diagnostic memory: dotnet-counters

* Running app
![image](https://user-images.githubusercontent.com/58478947/164983277-3bbd674e-23e4-4478-b944-c34bfd58b30c.png)

* Done calling api
![image](https://user-images.githubusercontent.com/58478947/164983328-7e94d26e-98b7-40ad-a114-c6b6ca98a3fd.png)

As you can see that memory still high like that after calling api for a while and GC does nothing?. Is this the way GC behaves ? Or is it me still not haven't optimized my app yet ?

**Comment by NguyenNguyen1995** [comment]
Here I added the `GC.Collect()`
![image](https://user-images.githubusercontent.com/58478947/164987955-7a6acac6-6e22-47a0-a58f-dd5b1bee54e1.png)

This is before I call api
![image](https://user-images.githubusercontent.com/58478947/164987995-d4f03d66-041e-4591-8d0c-61d5dd6f7414.png)

After calling
![image](https://user-images.githubusercontent.com/58478947/164988085-52ddaab6-0ab3-4811-93ee-ab7241ecebf3.png)

It's still there even I call `GC.Collect()`. According to my understand that we should not call `GC.Collect()` because the services container does for us by calling IDisposable , right?

**Comment by NguyenNguyen1995** [comment]
I saw this article mentioned about memory leak 
https://blog.markvincze.com/troubleshooting-high-memory-usage-with-asp-net-core-on-kubernetes/ 
Also when I turn on ServerGC is more worst than WorkstationGC (ServerGC runs nearly 3gb)

**Comment by davidfowl** [comment]
Do you want to know why the memory isn't being collected or do you want to improve the code so that it doesn't allocate this much memory?

**Comment by NguyenNguyen1995** [comment]
I turn on ServerGC in csproj
![image](https://user-images.githubusercontent.com/58478947/164988417-0c592525-ee82-4930-9139-3f14ac1c7056.png)

Before calling
![image](https://user-images.githubusercontent.com/58478947/164988445-d59e0f6b-af81-4387-b1e2-66165d0e7aa5.png)

During running 
![image](https://user-images.githubusercontent.com/58478947/164988500-05fd4c1b-e2c5-42f2-91a2-9d520e8d5d2c.png)

After calling
![image](https://user-images.githubusercontent.com/58478947/164988537-6d018e5e-7646-4898-bd1f-12d7b6a881e6.png)

**Comment by NguyenNguyen1995** [comment]
@davidfowl I want to know both. Thanks

**Comment by davidfowl** [comment]
For the first, get a memory profiler. There's one built into VS 2022 (if you have visual studio) https://docs.microsoft.com/en-us/visualstudio/profiling/memory-usage-without-debugging2?view=vs-2022:

![image](https://user-images.githubusercontent.com/95136/164988746-9c936035-668f-4748-b808-a07e18fe123d.png)

There are also other profiles you can use or you can capture a process dump and analyze it using `dotnet dump analyze`. There's a tutorial here https://docs.microsoft.com/en-us/dotnet/core/diagnostics/debug-memory-leak.

If this is your first time looking at a memory leak, these tools are more advanced and are harder to use.

**Comment by davidfowl** [comment]
For the second, you can load the 1M JSON objects from that file (how big is the file BTW) by using a FileStream instead of loading up a giant string in memory:

```C#
using var stream = new FileStream(dataPath, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 65536, useAsync: true);
await JsonSerializer.DeserializeAsync<List<Test>>(stream);
```

This will asynchronously load the JSON into Test objects using a internal 65K buffer. Instead of reading the entire file into memory before starting processing.

**Comment by NguyenNguyen1995** [comment]
I have run the `dotnet-dump` command and it looks like this
(it really long list so I just snap shot the last one)
![image](https://user-images.githubusercontent.com/58478947/164989251-e643b46c-0c94-4442-970d-e638e518370f.png)

**Comment by davidfowl** [comment]
Well you can see there's 300MB of strings and lots of EF overhead though I'm sure that's expected.

**Comment by NguyenNguyen1995** [comment]
I have run as you mentioned
Here is my configuration:
* TestService.cs
![image](https://user-images.githubusercontent.com/58478947/164989663-8d18b0ee-1e5c-4390-a0e0-02e17ab53760.png)

* demoDotnet.csproj
I have to turn off ServerGC due to high memory 
![image](https://user-images.githubusercontent.com/58478947/164989681-3d7b7610-b457-4f06-82d6-6aa8aec94100.png)

* HomeController.cs
I turn off the `GC.Collect()` call
![image](https://user-images.githubusercontent.com/58478947/164989732-9d037788-6567-4cd9-98aa-72cdee88735d.png)

It's still the same as It was
![image](https://user-images.githubusercontent.com/58478947/164989751-ee45528f-be46-4d5f-99c0-49033f12a5c4.png)

The result of dump file of test above
![image](https://user-images.githubusercontent.com/58478947/164989857-714ebba2-b8cb-4ebc-9a07-4972aae81e00.png)

**Comment by NguyenNguyen1995** [comment]
The one I highlight here seems cost a lot of memory
![image](https://user-images.githubusercontent.com/58478947/164989934-cb0abe88-e08c-4ff7-94e1-4f1258325018.png)

**Comment by roji** [comment]
As @davidfowl wrote above, EF Core does have a certain overhead. Assuming you're inserting a million rows, the above memory dump seems expected; we do intend to work on optimizing this in future releases, but there is nothing specifically problematic I see here. This memory should get released by the GC based on its own policies (depending on server vs. workstation, general memory pressure and other factorS).

**Comment by NguyenNguyen1995** [comment]
I see. Thanks

**Comment by davidfowl** [comment]
@NguyenNguyen1995 you should fix the string allocations.

**Comment by NguyenNguyen1995** [comment]
I don't know where should I fix it cause I only load data and then delegates to Add range  and save change, suddenly it log to my console all the insert statement. Does that cause string allocations high ?. I also try to call `ChangeTracker.Clear()`  but It seems allocations does not decrease.

**Comment by NguyenNguyen1995** [comment]
![image](https://user-images.githubusercontent.com/58478947/165003768-2da91dc0-9a1b-4c6e-b60e-c34576e55b11.png)
In this screenshot. I just load the demoDotnet.Test[] and nothing more.

**Comment by roji** [comment]
@NguyenNguyen1995 logging would indeed increase string allocations (and otherwise slow down your application). Verbose logging should generally not be turned on in production. Otherwise, see @davidfowl's suggestion above about handling JSON.

Re the GC... In your screenshot above, I see a commented-out GC.Collect in the controller method (GetAsync). At that point the context has not yet been disposed, and is still tracking all the data - even after SaveChanges has completed. Assuming your context is registered as scope, the moment the web request is completed (*after* the controller method returns), the context instance should get disposed, and at that point all the objects above become eligible for garbage collection. When exactly the GC will do that will vary.

**Comment by NguyenNguyen1995** [comment]
@roji I have implemented the way handling JSON as @davidfowl suggest and I appreciate. For logging, I don't know why it appear because usually if I insert a small amount (1-1000), it's not log anything but starting large records the console full of insert statement until it done saving. I have to turn entity framework core log to warning but still the string allocation is not decrease and I don't know where is coming from, I just load my data and that is

I've also read the doc and it's exactly as you said but why GC not collect such a big heap allocation like that, if my app constraint to only 1.2gb or in container with limited memory does GC collect them or OutOfMemoryException will throw ?

**Comment by roji** [comment]
Logging definitely should not be dependent on the number of or rows you insert. If you think you're seeing an issue with that, please open a new issue with a minimal, runnable code sample; please avoid posting screenshots - post an actual, runnable console program which we can run to investigate.

Regardless, you should not be getting an OutOfMemoryException - if there isn't enough memory, the GC should kick in and free any memory that isn't referenced any more (which should be the case above). I can't really answer more detailed questions on exact GC behavior, since those don't have anything to do with EF Core.

**Comment by yvshett** [comment]
@roji, Myself Working on a .NET Core 6, EF 7 on MySQL db. i use DI for all of my classes and dbContext. What i see is, my memory stays as is even after the api call ends. many time my memory exception i am getting. i see CPU % going above 300%. even though my app is allocated sufficient memory, still i see this issue.

**Comment by roji** [comment]
@yvshett we're not going to be able to help you without a minimal, runnable repro which shows these issues; it's simply not possible to know what you're doing from the very partial texutal description you're providing.
